cartographer_grpc/framework/retry.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <chrono>
+#include <thread>
+
+#include "cartographer_grpc/framework/retry.h"
+#include "glog/logging.h"
+
+namespace cartographer_grpc {
+namespace framework {
+
+RetryStrategy CreateRetryStrategy(RetryIndicator retry_indicator,
+                                  RetryDelayCalculator retry_delay_calculator) {
+  return [retry_indicator, retry_delay_calculator](int failed_attempts) {
+    if (!retry_indicator(failed_attempts)) {
+      return optional<Duration>();
+    }
+    return optional<Duration>(retry_delay_calculator(failed_attempts));
+  };
+}
+
+RetryIndicator CreateLimitedRetryIndicator(int max_attempts) {
+  return [max_attempts](int failed_attempts) {
+    if (failed_attempts >= max_attempts) {
+      return false;
+    }
+    return true;
+  };
+}
+
+RetryDelayCalculator CreateBackoffDelayCalculator(Duration min_delay,
+                                                  float backoff_factor) {
+  return [min_delay, backoff_factor](int failed_attempts) -> Duration {
+    CHECK_GE(failed_attempts, 0);
+    using cartographer::common::FromSeconds;
+    using cartographer::common::ToSeconds;
+    return FromSeconds(std::pow(backoff_factor, failed_attempts - 1) *
+                       ToSeconds(min_delay));
+  };
+}
+
+RetryStrategy CreateLimitedBackoffStrategy(Duration min_delay,
+                                           float backoff_factor,
+                                           int max_attempts) {
+  return CreateRetryStrategy(
+      CreateLimitedRetryIndicator(max_attempts),
+      CreateBackoffDelayCalculator(min_delay, backoff_factor));
+}
+
+bool RetryWithStrategy(std::function<bool()> op, RetryStrategy retry_strategy) {
```

Comment:
can these RetryWithStrategy() overloads be united in one? Perhaps, by providing a default reset function that does nothing.

Reply:
I had to check with the style guide, because I thought it is discouraged, but for non-virtual functions it's fine.